lib/lf/uscalfe/lagr_fe.h

```diff
@@ -710,6 +710,455 @@ class FeLagrangePoint : public ScalarReferenceFiniteElement<SCALAR> {
+   * and one to each edge of the triangle. There are no interior
+   * shape functions.
+   * @copydoc ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(dim_t codim) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 0) ? 0 : 1;
+  }
+
+  /**  @brief One shape function attached to each node
+   * and one to each edge of the triangle. There are no interior
+   * shape functions.
+   * @copydoc
+   * ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t,sub_idx_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      dim_t codim, sub_idx_t /*subidx*/) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 0) ? 0 : 1;
+  }
+
+  /** @copydoc ScalarReferenceFiniteElement::EvalReferenceShapeFunctions() */
+  [[nodiscard]] Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+  EvalReferenceShapeFunctions(const Eigen::MatrixXd& refcoords) const override {
+    LF_ASSERT_MSG(refcoords.rows() == 2,
+                  "Reference coordinates must be 2-vectors");
+
+    const size_type n_pts(refcoords.cols());
+    Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic> result(6, n_pts);
+
+    Eigen::ArrayXd x0 = refcoords.row(0).array();
+    Eigen::ArrayXd x1 = refcoords.row(1).array();
+
+    result.row(0) = (2.0 * (1 - x0 - x1) * (0.5 - x0 - x1)).matrix();
+    result.row(1) = (2.0 * x0 * (x0 - 0.5)).matrix();
+    result.row(2) = (2.0 * x1 * (x1 - 0.5)).matrix();
+    result.row(3) = (4.0 * (1 - x0 - x1) * x0).matrix();
+    result.row(4) = (4.0 * x0 * x1).matrix();
+    result.row(5) = (4.0 * (1 - x0 - x1) * x1).matrix();
+
+    return result;
+  }
+
+  /** @copydoc ScalarReferenceFiniteElement::GradientsReferenceShapeFunctions*/
+  [[nodiscard]] Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+  GradientsReferenceShapeFunctions(
+      const Eigen::MatrixXd& refcoords) const override {
+    LF_ASSERT_MSG(refcoords.rows() == 2,
+                  "Reference coordinates must be 2-vectors");
+
+    const size_type n_pts(refcoords.cols());
+    Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic> result(6, 2 * n_pts);
+
+    // reshape into a 12xn_pts matrix.
+    Eigen::Map<Eigen::Array<SCALAR, Eigen::Dynamic, Eigen::Dynamic>,
+               Eigen::AutoAlign>
+        temp(&result(0, 0), 12, n_pts);
+
+    Eigen::ArrayXd x0 = refcoords.row(0).array();
+    Eigen::ArrayXd x1 = refcoords.row(1).array();
+
+    // d/dx_0
+    temp.row(0) = -3.0 + 4.0 * x0 + 4.0 * x1;
+    temp.row(1) = 4.0 * x0 - 1.0;
+    temp.row(2) = 0.0;
+    temp.row(3) = 4 - 0 - 8.0 * x0 - 4.0 * x1;
+    temp.row(4) = 4.0 * x1;
+    temp.row(5) = -4.0 * x1;
+
+    // d/dx_1
+    temp.row(6) = -3.0 + 4.0 * x0 + 4.0 * x1;
+    temp.row(7) = 0.0;
+    temp.row(8) = 4.0 * x1 - 1.0;
+    temp.row(9) = -4 * x0;
+    temp.row(10) = 4.0 * x0;
+    temp.row(11) = 4.0 - 8.0 * x1 - 4.0 * x0;
+
+    return result;
+  }
```

Comment:
I suggest moving this implementation into `lagr_fe.cc` because it is quite long. For the first order element it's indeed very simple and the compile time overhead is almost negligible, but here its 30+ lines already.

Reply:
Thanks for your feedback!

 I have some questions regarding this remark:

1. As far as i have seen, currently all implementations of methods for templated classes are  written in the same header files as the classes. (e.g. for the currently implemented ElementMatrixProviders the eval method implementations are always written directly after the class definitions.). Should i do the same or indeed move the implementation to the cpp file.

2. If i separate this implementation into a separate cpp file, i  have to add lines of the form
`template void foo<double>();`
to avoid linking errors. For which types should I do this?

Reply:
ah, I missed that the class `FeLagrangeO2Tria` is a template, arrgh. Just forget about that comment and the next one, sorry for the confusion.

Reply:
(If you wanted to move it all out into a source files you would indeed have to add the lines of code that you mentioned :)